chips/sam4l/src/adc.rs

```diff
@@ -332,9 +354,21 @@ impl hil::adc::Adc for Adc {
                     return ReturnCode::FAIL;
                 }
             }
+
+            ReturnCode::SUCCESS
         }
+    }
+}
 
-        ReturnCode::SUCCESS
+/// Implements an ADC capable reading ADC samples on any channel.
+impl hil::adc::Adc for Adc {
+    type Channel = AdcChannel;
+
+    /// Enable and configure the ADC.
+    /// This can be called multiple times with no side effects.
+    fn initialize(&self) -> ReturnCode {
```

Comment:
I think there's a more fundamental question at play here, which is should we ever have an initialization function or should we expect the system to initialize at first use?

My opinion is that we should automatically initialize peripherals when we first go to use them, which allows the higher-level users to not have to keep track of initialization status and simplifies the interface layer.

I could see this breaking for devices with long initialization times (like the SD card, for instance). However, those types of peripherals need to have an asynchronous initialization function, with an initialization complete callback anyway.